Detectors/AOD/src/AODProducerWorkflowSpec.cxx

```diff
@@ -874,104 +873,68 @@ void AODProducerWorkflowDPL::run(ProcessingContext& pc)
-          auto& mcTruthTPC = tracksTPCMCTruth[trackIndex.getIndex()];
-          if (mcTruthTPC.isValid()) {
-            labelHolder.labelID = toStore.at(Triplet_t(mcTruthTPC.getSourceID(), mcTruthTPC.getEventID(), mcTruthTPC.getTrackID()));
-          }
-          if (mcTruthTPC.isFake()) {
-            labelHolder.labelMask |= (0x1 << 15);
-          }
-          if (mcTruthTPC.isNoise()) {
-            labelHolder.labelMask |= (0x1 << 14);
-          }
-          mcTrackLabelCursor(0,
-                             labelHolder.labelID,
-                             labelHolder.labelMask);
-        }
-        // its-tpc and its-tpc-tof labels
-        // todo:
-        //  probably need to store both its and tpc labels
-        //  for now filling only TPC label
-        if ((src == GIndex::Source::ITSTPC || src == GIndex::Source::ITSTPCTOF)) {
-          auto contributorsGID = recoData.getSingleDetectorRefs(trackIndex);
-          auto& mcTruthITS = tracksITSMCTruth[contributorsGID[GIndex::Source::ITS].getIndex()];
-          auto& mcTruthTPC = tracksTPCMCTruth[contributorsGID[GIndex::Source::TPC].getIndex()];
-          // its-contributor label
-          if (contributorsGID[GIndex::Source::ITS].isIndexSet()) {
-            if (mcTruthITS.isValid()) {
-              labelHolder.labelITS = toStore.at(Triplet_t(mcTruthITS.getSourceID(), mcTruthITS.getEventID(), mcTruthITS.getTrackID()));
+        if (GIndex::includesSource(src, mInputSources)) {
+          auto mcTruth = recoData.getTrackMCLabel(trackIndex);
+          MCLabels labelHolder;
+          if (src == GIndex::Source::MFT) { // treating mft labels separately
+            if (mcTruth.isValid()) {        // if not set, -1 will be stored
+              labelHolder.labelID = mToStore.at(Triplet_t(mcTruth.getSourceID(), mcTruth.getEventID(), mcTruth.getTrackID()));
             }
-          }
-          if (contributorsGID[GIndex::Source::TPC].isIndexSet()) {
-            if (mcTruthTPC.isValid()) {
-              labelHolder.labelTPC = toStore.at(Triplet_t(mcTruthTPC.getSourceID(), mcTruthTPC.getEventID(), mcTruthTPC.getTrackID()));
+            if (mcTruth.isFake()) {
+              labelHolder.mftLabelMask |= (0x1 << 7);
             }
+            if (mcTruth.isNoise()) {
+              labelHolder.mftLabelMask |= (0x1 << 6);
+            }
+            mcMFTTrackLabelCursor(0,
+                                  labelHolder.labelID,
+                                  labelHolder.mftLabelMask);
+          } else {
+            if (mcTruth.isValid()) { // if not set, -1 will be stored
+              labelHolder.labelID = mToStore.at(Triplet_t(mcTruth.getSourceID(), mcTruth.getEventID(), mcTruth.getTrackID()));
+            }
+            // treating possible mismatches for global tracks
+            auto contributorsGID = recoData.getSingleDetectorRefs(trackIndex);
+            if (contributorsGID[GIndex::Source::ITS].isIndexSet() && contributorsGID[GIndex::Source::TPC].isIndexSet()) {
```

Comment:
Also here the reason of the `&&` is not clear. For the record: the MCCompLabel of standalone detector track (ITS, TPC, TRD tracklets collection etc) is set to refer to the dominant particle represented in the contributors (e.g. clusters) of the track and `setFake()` is set if the fraction of contributors w/o this dominant particle is above threshold (normally 10% for TPC, >0 for other detectors).
Then, for the compound tracks labels the rule is: it refers on the particle mentioned in the TPC track label and the isFake is set if the track/cluster in other detector matched to TPC has not this TPC label. I.e. the fakeness of individual tracks is ignored on this level.

So, if you have e.g. ITS-TPC-TOF match contributed by
ITStrack:  {mcIDA, non-fake}
TPCtrack: {mcIDB, fake}
TOCclus: {mcIDB}
then the label of ITS-TPC-TOF will be {mcIDB, non-fake}, it will refer to  track ITS-TPC with label {mcIDB, fake} which in turn will refer to ITS track  {mcIDA, non-fake} and TPCtrack  {mcIDB, non-fake}.

Reply:
Hi @shahor02. I added this additional check to represent possible ITS and TPC label mismatches for global tracks that have both ITS and TPC labels -- so I put `&&`. Do I understand correctly that such check is not needed, since `isFake()` can be used to indicate any mismatches?

Reply:
Hi @shahor02. I added this additional check to represent possible ITS and TPC label mismatches in label masks for global tracks that have both ITS and TPC labels -- so I put &&.
https://github.com/nburmaso/AliceO2/blob/ab000dee139c59213711f14aaa15541942afd1d7/Detectors/AOD/src/AODProducerWorkflowSpec.cxx#L917-L920
Do I understand correctly that such check is not needed, since isFake() can be used to indicate any mismatches?

Reply:
Yes, for compound tracks (TPC always there) the label MC particle ID points to that of the TPC track label and isFake is set in case there is a mismatch between the MC particle IDs of the 2 contributors (as on schema below) of this global track
````
{ITS-TPC-TRD-TOF}
       ^
{ITS-TPC-TRD}, {TOF}
       ^
{ITS-TPC}, {TRD}
       ^
{ITS}, {TPC}
````

````
{ITS-TPC-TOF}
       ^
{ITS-TPC}, {TOF}
````